### PR TITLE
Extend XML::Reader with more LibXML methods

### PR DESCRIPTION
**Pull request by felixbuenemann**
This adds a couple of method bindings that come in handy when doing pull parsing or hybrid parsing (search with pull then expand node).

### CODE REVIEW

File: src/xml/reader.cr [hunk 1]
```diff
@@ -25,6 +25,14 @@ struct XML::Reader
     LibXML.xmlTextReaderRead(@reader) == 1
   end
 
+  def next
+    LibXML.xmlTextReaderNext(@reader) == 1
+  end
+
+  def next_sibling
```
Comment (RX14) [inline src/xml/reader.cr]: libXML docs say this and `next` are the same, are they? More docs in general would be great.

**Comment by RX14** [comment]
I'm fine with this but I think the new methods should have docs and specs. Even if the surrounding methods don't have docs.

If you want to document the whole class though, that would be fantastic :)

File: src/xml/reader.cr (hunk 1 above)
Comment (felixbuenemann) [reply src/xml/reader.cr]: No, they are not the same.

See this example:

```
<root>
  <childa/>
  <childb/><!-- current reader node -->
</root>
```

If you call `reader.next` would return `true` and it will move to `</root>`, but `reader.next_sibling` would return `false`, because there is no other sibling under the root node.

See:
* [xmlTextReaderNextSibling](https://github.com/GNOME/libxml2/blob/master/xmlreader.c#L2037)
* [xmlTextReaderNext](https://github.com/GNOME/libxml2/blob/master/xmlreader.c#L1669)

That's the theory at least, the system `/usr/lib/libxml2.2.dylib` on my mac which should be 2.9.4 according to `xml2-config --version` always returns `-1`, looks like I'll have to recompile crystal against the latest version from homebrew to see if it is a bug in that specifc version.

Comment (felixbuenemann) [reply src/xml/reader.cr]: Yeah, that appears to be a bug in libxml2, it works after applying the following patch to master:

```diff
diff --git a/xmlreader.c b/xmlreader.c
index 4053269b..c195d875 100644
--- a/xmlreader.c
+++ b/xmlreader.c
@@ -2037,7 +2037,7 @@ int
 xmlTextReaderNextSibling(xmlTextReaderPtr reader) {
     if (reader == NULL)
         return(-1);
-    if (reader->doc == NULL) {
+    if (reader->doc != NULL) {
         /* TODO */
        return(-1);
     }
```

I created GNOME/libxml2#13 to get this fixed in libxml.

**Comment by felixbuenemann** [comment]
Yeah, none of the `XML::Reader` class is currently tested.

I could copy over the comments from the libxml docs and adjust for the slightly different return values.

**Comment by RX14** [comment]
If the function is broken on all released versions of libxml2, what's the point in adding it? Adding in a function that we *know* is going to be buggy for many years to come because of out-of-date distros is just going to be painful. Can we work around this bug in crystal?

**Comment by felixbuenemann** [comment]
We could implement our own version of the next_sibling method. I think I'll have to add some more bindings for that, but it should work.

File: src/xml/reader.cr (hunk 1 above)
Comment (felixbuenemann) [reply src/xml/reader.cr]: I have pushed a workaround that makes `next_sibling` work even with the incomplete `xmlTextReaderNextSibling()`.

**Note:** The libxml2 patch posted above is outdated and broken, see GNOME/libxml2#13 for the revised version.

**Comment by felixbuenemann** [comment]
I'll start working on specs for `XML::Reader` next…

**Comment by felixbuenemann** [comment]
@RX14 Please review.

As suggested in your earlier review I've documented all the methods in `XML::Reader` and added specs.

I've also fixed a few methods that could crash, if called before the first or after the last call to `XML::Reader#read`, but as a consequence those methods can now return `nil`.

Would it be better to adjust the code so that they raised `XML::Error` instead?

That would reduce the number of nil checks required when working with them and with proper usage the `nil` case should never be encountered.

**Comment by RX14** [comment]
> I've also fixed a few methods that could crash, if called before the first or after the last call to XML::Reader#read, but as a consequence those methods can now return nil.

I think that should be a separate PR, since it's a breaking change.

**Comment by felixbuenemann** [comment]
>> I've also fixed a few methods that could crash, if called before the first or after the last call to XML::Reader#read, but as a consequence those methods can now return nil.

> I think that should be a separate PR, since it's a breaking change. Or actually - if they currently segfault in this condition, then it's not really a breaking change to make them raise.

I changed the getter methods that could return either a String or Nil to always return a String. In the cases where nil was returned previosuly, I now return an empty string. This avoids breaking backwards compatibility and makes more sense than raising an error.

I also renamed the `XML::Reader#attribute` method to `XML::Reader#[]?` and added `#[]` which raises `KeyError` if the attribute is not found.

**Comment by RX14** [comment]
Returning an empty string is unacceptable. Lets revert all the changes to behaviour in other methods, and keep this PR focussed on docs and adding *new* methods. We can discuss those methods in another PR.

**Comment by felixbuenemann** [comment]
Oh common, unacceptable?

I had that idea by looking at the API of XML::Node which does exactly that, see:

* [XML::Node#content](https://github.com/crystal-lang/crystal/blob/master/src/xml/node.cr#L94) (equivalent to XML::Reader#value)
* [XML::Node#name](https://github.com/crystal-lang/crystal/blob/master/src/xml/node.cr#L256) (equivalent to XML::Reader#name)

So please enlighten me why you think it is unacceptable in this case?

I think it is a good idea since it avoids nil checks in user code just to handle one edge case.

**Comment by felixbuenemann** [comment]
Let's recap why I chose empty strings instead of raising:

* `XML::Reader#name` only returns a null pointer, when it is not currently on a node, so this is before the first or after the last call to `#read`. This method will only return an empty string in this edge case, in other cases it will either return the element name, or something like `"#text"` or `"#comment"`. This means the edge case can be detected by checking for en empty string instead of having to clutter the code with nil checks just to handle the edge case that doesn't happen if the api is properly used.
* `XML::Reader#value` returns a null pointer in two cases: when the node doesn't have text content, eg `<root/>` or before the first or after the last `#read`. Since we can't differentiate between these two states, it would be a very bad idea to raise here, since you might want to parse xml like `<root><child/><child>text</child></root>` which means you would have to add additional checks to `#empty_element?` just to ensure you don't run into exception, while returning an empty string for "empty content" seems perfectly fine for the majority of use cases and if the user really cares if the node has no text content he can just as well call `#empty_element?`.

So I urge you to reconsider your point of view of allowing empty strings in these methods.

**Comment by straight-shoota** [comment]
Please slice it up already. It's too confusing to discuss several topics in one place. Let's have one discussion about adding new stuff and one about changing existing methods.

**Comment by felixbuenemann** [comment]
I think special casing `#name` and `#value` to raise an error, while all the other existing `XML::Reader` methods silenty swallow errors is not a good idea, since it makes the API inconsistent.

There are a lot of methods that do something like `if LibXML.xmlFoo(@reader) == 1` and they all ignore that those methods can return `-1` and treat it as `false` instead of `nil` or an exception.

I think either all reader methods should raise on error or none.

The way you usually use the reader api is like this:

```crystal
reader = XML::Reader.new(io)
while reader.read
  # do stuff
end
```

So you will never hit those edge cases unless you use the api in a completely wrong way.

**Comment by felixbuenemann** [comment]
@straight-shoota Look, I added tests for the entire `LibXML::Reader` here and documented the methods, as suggested by @RX14 and +1ed by you at the beginning of this PR. This revealed that two methods could segfault when the reader was in an invalid state due to missing NULL pointer checks, so I fixed them making the tests pass, because having a PR with failing tests would be a bad idea.

If the entires API of `XML::Reader` should be changed to raise if any of the LibXML methods return an error, that is a huge breaking change and should be tackled in a separate PR.

**Comment by felixbuenemann** [comment]
@RX14 I think your review is stale. Can this be merged?

**Comment by RX14** [comment]
Lets just rip out `XML` entirely. Put it in a shard. It's not mature enough to be in the stdlib, as these revelations prove.

**Comment by RX14** [comment]
For now, this can be merged so the fixes are in if/when we split XML into a shard.

**Comment by felixbuenemann** [comment]
@RX14 I'm unopinionated wether this should be in stdlib or not, but if you choose to extract it into a shard feel free to ping me – I might be able to help with maintenance.
